### DISCUSSION TITLE
Sharing a USB camera with usbipd in a docker container

### DISCUSSION THREAD

**Discussion by guillaumenibert** [microsoft/WSL#11230] [Q&A]
Hello everyone, thank you for your work, I am looking to share my USB camera to a docker container.

I am not sure this is a WSL issue, so I am posting my problem here.

### Windows version

Microsoft Windows [version `10.0.22631.3155`]

### WSL Version

2.0.14

### Are you using WSL 1 or WSL 2?

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

WSL kernel compiled with USB Camera drivers based on version `5.15.133.1`; Tutorial provided by @AgileDevArt available here: https://youtu.be/t_YnACEPmrM?feature=shared&t=183.

### Distro Version

Docker Desktop for Windows `v4.28.0`; `docker-desktop` distro.

### Other Software

usbipd-win `v4.1.1-dev` from https://github.com/dorssel/usbipd-win/issues/856#issuecomment-1931850093.

### Repro Steps

Based on discussion https://github.com/dorssel/usbipd-win/issues/315, I performed these steps:
 1. I recompiled the Linux kernel for WSL with camera driver support,
 2. then attached my camera to WSL using usbipd,
 3. tested the camera on the Ubuntu default WSL distribution with `cheese`: **it works**.
 4. `/dev/video0` is visible in the `docker-desktop` distro.
 5. I tried this tip https://github.com/dorssel/usbipd-win/issues/315#issuecomment-1075315900 from @NikLeberg :

> Hi @salvatoreTorsello I use usbipd successfully to forward a USB device to a docker container. For that I had to map `-v /dev:/dev` and add `--privileged` so the container can access the devices. Depending on you device you have to install drivers for it in WSL2 thought.

Adapting it, I tried these configurations (docker compose file), with or without privileges, with or without the *devices* part,

```yml
...
privileged: true
volumes:
   - /dev:/dev
   - /run/desktop/dev:/dev
   - /run/desktop/dev/video0:/dev/video0

devices:
   - /dev:/dev
   - /run/desktop/dev:/dev
   - /dev/video0:/dev/video0
...
```

### Expected behaviour

I should see `/dev/video0` in my container.

### Actual behaviour

`/dev/video0` is visible in the `docker-desktop` distro, but not in the container: when I do `lsusb` in my container, the camera is visible, but the output of `ls /dev` does not list `video0`.

### Diagnostic logs

*No response*

### Additional information

*Copy of this post in the Docker for Windows GitHub repository: https://github.com/docker/for-win/issues/13940 and in the usbipd project discussions: https://github.com/dorssel/usbipd-win/discussions/879; @dorssel advised me to also post my problem to WSL's GitHub repository.*

---
I am interested in any advice, whether or not it concerns WSL or rather Docker Desktop for Windows.

Best regards,

Guillaume Nibert

**Comment by PompIsaac123** [comment]
did you figure it out? im triying to do the same with wsl2. probably just better to not use docker ? idk. looking for answers.

**Reply by guillaumenibert** [reply]
Hello,

Here is what I did to make it work (https://github.com/docker/for-win/issues/13940#issuecomment-2009725125):

> > @guillaumenibert any updates on this?
> 
> I have tried again, and it seems that rebooting the computer has made `/dev/video0` visible. Unfortunately, despite the presence of `/dev/video0` in the container I cannot access the device. I have tried :
> 
>     * setting the container to be `privileged`
> 
>     * adding my container user to the `video` group
> 
> 
> It does not work.
> 
> The only solution I have found which works is to enter the `docker-desktop` system
> 
> ```powershell
> wsl --distribution docker-desktop
> ```
> 
> and change the owner of `/dev/video0` to one with `UID:GID` = `1000:1000`
> 
> ```shell
> chown 1000:1000 /dev/video0
> ```
> 
> The container now has access to the host's camera.
> 
> Best regards,
> 
> Guillaume Nibert
